2019-08-03-q-and-as-about-nix-for-haskellers.md

```diff
@@ -0,0 +1,692 @@
+
+    All of these groups of people have their own code in nixpkgs that others
+    often don't venture into.  For example, I've been using Nix for
+    about a year and a half, but I don't really have any idea how the
+    Python libraries are structured, how cross compiling works, etc.
+
+    If I wanted to play around with cross compiling, it would be time-consuming
+    to come up to speed with how everything works.
+
+Learning Nix is a non-trivial endeavour.  However, Nix is extremely flexible.
+Like Haskell, I'd say the time-investment can be worth it depending on how
+extensively you're planning on using it.
+
+With the above knowledge, it should be possible to dive into the questions from
+Dmitrii.
+
+## Questions and Answers
+
+Below, I list Dmitrii's questions and answer them one-by-one.[^7]  These answers
+should be understandable by people with no experience using Nix.
+
+### 1. What is the difference between `nix-build` and `nix-shell`?
+
+`nix-build` and `nix-shell` are the two most widely used Nix tools.
+
+They both evaluate derivations in a similar way, but their final results are
+different.
+
+When `nix-build` is finished evaluating a derivation, it produces a directory
+in the Nix store containing the binaries, shared objects, header files, etc for
+the derivation.
+
+When `nix-shell` is finished evaluating a derivation, it launches a Bash shell.
+It makes all the inputs to the derivation available in the Bash shell.  For
+instance, imagine you had a derivation that depends on `git`.  If you ran
+`nix-shell` on that derivation, it would put you in a Bash shell with an
+environment where the `git` CLI tools available.  From `nix-shell`,
+it should be possible to build the derivation manually (running commands like
+`./configure && make && make install`).
+
+By default, if you give no arguments to `nix-build`, it looks for a file called
+`default.nix` in the current working directory.  It assumes this file defines a
+derivation.  It builds this derivation.
+
+If you give no arguments to `nix-shell`, it looks for a file called `shell.nix`
+in the current working directory.  It assumes this files defines a derivation.
+It takes all the inputs from this derivation and launches a Bash shell with all
+these inputs available.
+
+In later questions, I talk about how `nix-build` and `nix-shell` are related to
+doing Haskell development.
+
+### 2. When would somebody want to use Nix instead of cabal or stack?
+
+In general, there are three ways to use Nix to build Haskell projects: using
+`nix-build` directly, using `cabal` + Nix, or using `stack` + Nix.
+
+But before I describe each of these methods, I need to talk about Haskell
+package sets in Nix.
+
+As explained in the previous section, the Nix language provides a way to define
+a _derivation_. A derivation describes the build steps for building a library,
+a binary, etc.
+
+The nixpkgs repo provides functions that make it easy to define a derivation
+for a Haskell package.
```

Comment:
I find the last three paragraphs hard to read. Maybe it's worth expanding on them. For example, from my point of view, as a Haskell developer, I see the rest of the world divided in two parts: Haskell packages which I use by cabal/stack and everything else which I had to use some arcane unknown tools to install. But since `nix` unifies this approach, all such packages are the same (if I correct). So Haskell packages in the nix ecosystem is just a list of packages under specified namespace. Of course, only if this is correct :slightly_smiling_face: